architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+## The Core Problem
+
+The core problem is that CRDs (being a globally shared resource) are fragile. Once a CRD is installed, we typically have to assume (all other things being equal) that it is shared across namespaces and groups of users.
+
+For that reason, installing, modifying, and deleting CRDs is a process that has ramifications for all users and systems of that cluster.
+
+In spite of this, the Kubernetes API server is permissive about these CRDs. CRDs are mutable (even without a version change). When they are deleted, all instances of them are deleted without warning. They can be upgraded to be entirely incompatible with previous versions. And there is no way to programmatically inspect the CRDs in a cluster and determine whether they are used, how they are used, and by what.
+
+## Users First: A Core Principal
+
+Over the years, several proposals have surfaced and been rejected for one straightforward reason: They did not protect the user from badly written charts. This section explains the reasoning process behind our decision-making.
+
+Helm distinguishes between at least two roles:
+
+- A chart author: A person filling this role _creates_ and _maintains_ Helm charts
+- A Helm user: This person installs and manages instances of charts
+
+### The Chart Author Role
+
+We assume that a _chart author_ has three specific areas of domain knowledge:
+
+1. How to model applications in Kubernetes
+2. How to create Helm charts, including author templates
+3. How the application they are packaging works
+
+In this case, we assume the chart author role includes knowledge of Kubernetes kinds, Helm template commands, and so on.
+
+We do not assume that chart authors have knowledge about the clusters into which their charts are deployed, knowledge of the source code for the packages they install, or knowledge of the extended toolchains that their users have employed. Furthermore, we do not assume that chart authors will always follow best practices or accommodate use cases that may be important to some class of users. In fact, the Helm security model urges us to include bad actors in the class of chart authors. (That is, there is or may be a small subclass of chart authors that have intentions counter to the desires of their target Helm users.)
+
+### The Helm User Role
+
+Our assumptions about _Helm users_ are more modest. We do not assume they know much about Kubernetes or Helm--perhaps only enough to follow the Quickstart guide for Helm. With the base Helm user, we do not assume that they know what a Pod is, let alone a CRD. While we do assume that they know a little about YAML, we make no assumptions that they know about the Kubernetes flavor of YAML.
```

Comment:
We probably need to assume that a _Helm user_ may not even have permission to install cluster-wide resources like CRDs or ClusterRoles. This is relevant to the discussions later about the complications introduced by multi-tenant clusters, which this is more-likely to be the case.

Reply:
Sometimes people have permission and sometimes they don't. In many cases, they should not have permission. For that situation, Helm really needs a way to easily extract CRDs for someone with permission to install.

Reply:
I just created https://github.com/helm/helm/issues/8607